### ISSUE TITLE
Classification metrics docs incorrectly state they work with logits

### ISSUE THREAD

**Issue by timlod** [PyTorchLightning/pytorch-lightning#5879]
Currently, all classification metrics (ie. https://pytorch-lightning.readthedocs.io/en/stable/metrics.html#accuracy) state:
`Accepts logits from a model output or integer class values in prediction.`.

In the parameter section, it states:
`threshold
(float) – Threshold value for binary or multi-label logits. default: 0.5`

This suggests that predictions have already passed through a sigmoid at this stage, which logits have not.
The default threshold of 0.5 does not make sense with logits, or more correctly, raw model predictions (before sigmoid/softmax).

In #5340, the code around Accuracy was changed to say `preds` instead of `logits`, but the documentation remains to state `logits`.

I see two options (explained in the case of binary classification):
1. Continue accepting logits, but making explicit that inside the metric either the predictions will be passed through a sigmoid, or the threshold will be passed through a logit function [it is important that the documentation makes explicit that the given threshold still needs to be between 0 and 1 then].
2. The documentation changes from logits to probabilities, and explicitly states that one needs to apply sigmoid/softmax before passing predictions into the metric (this will be the case when users use `CrossEntropyLoss` or `BCEWithLogitsLoss`)

Personally, I would prefer a solution that accepts a parameter that can work with raw model outputs, as that would work well together with the more efficient `BCEWithLogitsLoss`/`CrossEntropyLoss`. This could also be a flag, like `is_logit=False`, which would apply sigmoid/softmax when True. That would also be required if one wants to be able to interpret thresholds returned by `PrecisionRecallCurve`.

**Comment by SkafteNicki** [comment]
Hi @timlod,
So we are aware that many classification metrics have stated that they support logits when the actually do not. Many of the classification metrics have been overhaul since v1.1 and for example accuracy metric does not mention logits anymore (see the following file on the v1.2 branch):
https://github.com/PyTorchLightning/pytorch-lightning/blob/release/1.2-dev/pytorch_lightning/metrics/classification/accuracy.py
Moving forward classification metric (for now) will only support integer labels or probabilities. 

That said, I can see that some metrics still need to be updated with `logits`->`preds` like confusion matrix:
https://github.com/PyTorchLightning/pytorch-lightning/blob/release/1.2-dev/pytorch_lightning/metrics/classification/confusion_matrix.py

**Comment by timlod** [comment]
Hi @SkafteNicki,
Thanks for the quick response.
Do you think we could add the logit functionality through a flag, as described in my option 1.)?
If possible, one should use `BCEWithLogitsLoss` and `CrossEntropyLoss` for efficiency/numerical stability reasons.

Right now, I believe I have to do something like this:
```
if n_classes == 1:
    pred = torch.sigmoid(pred)
else:
    pred = torch.softmax(pred, 1)
classification_metrics(pred, y)
```
(`classification_metrics` calls `metric(pred, y)` for each metric)
to get correct metrics. (It is exacerbated by the fact that my architecture needs to work with arbitrary outputs/losses)

**Comment by SkafteNicki** [comment]
I am not against adding support for logits. Advantage: more freedom to the user. Disadvantage: additional flag in metric construction. For now I am going to create a PR that fixes the documentation, so it is clear that we (again, for now) expects probabilities and not logits.
I would argue that you should still be able to use `BCEWithLogitsLoss` and `CrossEntropyLoss` for efficiency/numerical stability as you can just convert to labels directly for metric computations:
``` python
if n_classes = 1:
    pred = pred > 0
else:
    pred = pred.argmax(dim=1)
```
maybe @tadejsv have an something to say about this (how easy would it be to extend current interface to logits)?

**Comment by timlod** [comment]
I'm making heavy use of `PrecisionRecallCurve`, which requires probabilities directly. Hence I can't circumvent (part of) the problem by getting labels directly :(.
I agree it's better for now to expect probabilities. I'd +1 logits support - I may be able to help with the actual implementation in form of a PR as well if it's not a huge piece of work.

**Comment by SkafteNicki** [comment]
@timlod when I look at the `PrecisionRecallCurve` metric I do not see anything that prevents you from using logits?

**Comment by timlod** [comment]
You're right, it should be possible to compute the metric from metrics. However, the thresholds it returns are in terms of logit, and again need to be transformed with a sigmoid. (When plotting the curve, I visually indicate interesting thresholds)

**Comment by SkafteNicki** [comment]
Documentation is now updated by PR #5880. Closing this issue.
@timlod feel free to send a PR with improvements to a single or multiple metrics with support for logits.
